### PR TITLE
VZ-5716 Install and Upgrade implementation for Arbirary Helm Values

### PR DESCRIPTION
**Pull request by davisgomes**
# Description

This is the design for the integration of arbitrary helm values in the Install and Upgrade process. Changes to the internal structure had to be made to support precedence.

Fixes VZ-5716

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 1]
```diff
@@ -354,25 +351,28 @@ func (h HelmComponent) Reconcile(_ spi.ComponentContext) error {
...
-func (h HelmComponent) buildCustomHelmOverrides(context spi.ComponentContext, namespace string, additionalValues ...bom.KeyValue) (helm.HelmOverrides, error) {
+func (h HelmComponent) buildCustomHelmOverrides(context spi.ComponentContext, namespace string, additionalValues ...bom.KeyValue) ([]helm.HelmOverrides, error) {
 	// Optionally create a second override file.  This will contain both image setOverrides and any additional
 	// setOverrides required by a component.
 	// Get image setOverrides unless opt out
-	overrides := helm.HelmOverrides{}
 	var kvs []bom.KeyValue
 	var err error
+
+	// TODO: get the values from the override list and populate helm override values
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:361]: This needs to be done at component level I suppose?

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:367]: No, I think we want to do this at the helm component level

File: pkg/helm/helmcli.go [hunk 2]
```diff
@@ -33,10 +33,10 @@ type ChartStatusFnType func(releaseName string, namespace string) (string, error
...
 // HelmOverrides contains all of the overrides that gets passed to the helm cli runner
 type HelmOverrides struct {
-	SetOverrides       string   // for --set
-	SetStringOverrides string   // for --set-string
-	SetFileOverrides   string   // for --set-file
-	FileOverrides      []string // for -f
+	SetOverrides       string // for --set
+	SetStringOverrides string // for --set-string
+	SetFileOverrides   string // for --set-file
+	FileOverride       string // for -f
```
Comment (kminder) [inline pkg/helm/helmcli.go:39]: I'll read more but does this imply we are merging the various helm value sources instead of letting helm do that?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 3]
```diff
@@ -75,6 +75,9 @@ const (
 
 	// InstancePrincipal is used for instance principle auth type
 	instancePrincipal authenticationType = "instance_principal"
+
+	extraArgsKey  = "extraArgs[0]"
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:79]: How are these changes related?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 4]
```diff
@@ -351,6 +354,23 @@ func isOCIDNS(vz *vzapi.Verrazzano) bool {
 
 // AppendOverrides Build the set of cert-manager overrides for the helm install
 func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ string, kvs []bom.KeyValue) ([]bom.KeyValue, error) {
+	// use image value for arg override
+	bomFile, err := bom.NewBom(config.GetDefaultBOMFilePath())
+	if err != nil {
+		return kvs, compContext.Log().ErrorNewErr("Failed to get the bom file for the Prometheus Operator image overrides: ", err)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:360]: BOM

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 5]
```diff
@@ -61,6 +64,9 @@ type HelmComponent struct {
 	// AppendOverridesFunc is an optional function get additional override values
 	AppendOverridesFunc appendOverridesSig
 
+	// SupportsHelmValueOverrides Indicates whether or not the component supports helm value overrides
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:67]: Fix comment to match

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 6]
```diff
@@ -226,6 +235,7 @@ func (h HelmComponent) Install(context spi.ComponentContext) error {
 
 	// vz-specific chart overrides file
 	overrides, err := h.buildCustomHelmOverrides(context, resolvedNamespace, kvs...)
+	defer vzos.RemoveTempFiles(context.Log().GetZapLogger(), `\w*`)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:238]: Is this deletion scoped well enough to ensure unrelated temp files are not deleted?

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 7]
```diff
@@ -389,43 +419,156 @@ func (h HelmComponent) buildCustomHelmOverrides(context spi.ComponentContext, na
...
+		yamlValues = append(yamlValues, yamlValue)
+	}
+
+	context.Log().Infof("YAML Values: %v", yamlValues)
+
+	// Take the yaml values and construct a yaml file
+	// Each value is overalyed by the next value, and lists are replaced
+	fileString, err := yaml.ReplacementMerge(yamlValues...)
+	if err != nil {
+		return newKvs, context.Log().ErrorfNewErr("Failed to convert yaml values to a yaml file: %v", err)
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:453]: YAML

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 8]
```diff
@@ -389,43 +419,156 @@ func (h HelmComponent) buildCustomHelmOverrides(context spi.ComponentContext, na
...
+func createHelmOverrideFile(ctx spi.ComponentContext, nsn types.NamespacedName, data map[string]string, dataKey string, optional *bool) (*os.File, error) {
+	var file *os.File
+
+	// Get resource data
+	fieldData, ok := data[dataKey]
+	if !ok {
+		if optional == nil || !*optional {
+			err := ctx.Log().ErrorfNewErr("Could not get Data field %s from Resource %s from namespace %s", dataKey, nsn.Name, nsn.Namespace)
+			return file, err
+		}
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:561]: Are we completely  honouring the case where optional is set to true? Like the name field is optional, what if the name is not supplied but the key is? Maybe the code should check all the resources for that key if optional is true, and if it's false enforce that resource name must be supplied?

File: platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go [hunk 9]
```diff
@@ -68,3 +69,19 @@ func (c prometheusComponent) IsReady(ctx spi.ComponentContext) bool {
...
 }
+
+// ValidateInstall verifies the installation of the Verrazzano object
+func (c prometheusComponent) ValidateInstall(effectiveCR *vzapi.Verrazzano) error {
+	if effectiveCR.Spec.Components.PrometheusOperator != nil {
+		return vzapi.ValidateHelmValueOverrides(effectiveCR.Spec.Components.PrometheusOperator.ValueOverrides)
+	}
+	return nil
+}
+
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:80]: Is this function being used anywhere?

File: platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go [hunk 10]
```diff
@@ -68,3 +69,19 @@ func (c prometheusComponent) IsReady(ctx spi.ComponentContext) bool {
 func (c prometheusComponent) PreInstall(ctx spi.ComponentContext) error {
 	return preInstall(ctx)
 }
+
+// ValidateInstall verifies the installation of the Verrazzano object
+func (c prometheusComponent) ValidateInstall(effectiveCR *vzapi.Verrazzano) error {
```
Comment (adiforluls) [inline platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:74]: Is this being used anywhere?

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 3 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:79]: Because I had to convert Yaml values to files, the way that this argument was separated broke my  yaml converter. Before this change, it looked like this:
```
extraArgs[0]--argument:
  image-name:v1.0
```

Now it formats to:
```
extraArgs:
  - --argument=image-name:v1.0
```

File: platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go [hunk 11]
```diff
@@ -68,3 +69,19 @@ func (c prometheusComponent) IsReady(ctx spi.ComponentContext) bool {
...
+	return nil
+}
+
+// ValidateUpgrade verifies the upgrade of the Verrazzano object
+func (c prometheusComponent) ValidateUpgrade(effectiveCR *vzapi.Verrazzano) error {
+	if effectiveCR.Spec.Components.PrometheusOperator != nil {
+		return vzapi.ValidateHelmValueOverrides(effectiveCR.Spec.Components.PrometheusOperator.ValueOverrides)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:87]: No. This same function is used for both install and upgrade because the values come from the same source.

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go (hunk 8 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:561]: If the name and key pair are not located, the optional field allows it to successfully skip that key value pair without error. Similarly, if a name is found without a key, the upgrade should pass without error.

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go (hunk 6 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:238]: To my knowledge, this doesn't interfere with any other temp files. It doesn't seem that any higher level temp files get created before and used after this function.

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go (hunk 4 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go]: No, I updated the comment to match the component

File: platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (adiforluls) [reply platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:87]: Ah, what I meant was maybe we need something like a validateComponentOverride? Although if we're gonna do a component level validation of values in the future, this pattern seems to stick

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go (hunk 8 above)
Comment (adiforluls) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:561]: What about this: key is found without a name? I think key isn't optional so will have to be supplied no matter what

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:561]: The key will not be found without a name. The ConfigMap search is done first. If no ConfigMap with the given name is found, we give up.

File: platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:87]: This function already existed in the Helm component, so it seemed like a good fit.

[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/prometheus/operator/promoperator_component.go:87]: I understand your question, and I had the same thought. I thought it would be an unnecessary abstraction for a very small amount of code.

**Review by adiforluls — APPROVED** [review]

**Comment by adiforluls** [comment]
LGTM

**Review by stevshap — APPROVED** [review]

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go [hunk 12]
```diff
@@ -740,21 +740,11 @@ type OciLoggingConfiguration struct {
 }
 
 type HelmValueOverrides struct {
-	MonitorChanges *WatchHelmValues `json:"monitorChanges,omitempty"`
-	ValueOverrides []Overrides      `json:"overrides,omitempty"`
+	MonitorChanges *bool       `json:"monitorChanges,omitempty"`
```
Comment (michael-cico) [inline platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:743]: I realize this is late to the game, but should this just be called `ValueOverrides`?  There are components like `Istio` that might use this that don't use Helm.

Not for this PR of course.

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 13]
```diff
@@ -389,43 +419,156 @@ func (h HelmComponent) buildCustomHelmOverrides(context spi.ComponentContext, na
...
-		for _, kv := range kvs {
-			if kv.SetString {
-				if setStringOverridesBldr.Len() > 0 {
-					setStringOverridesBldr.WriteString(",")
-				}
-				setStringOverridesBldr.WriteString(fmt.Sprintf("%s=%s", kv.Key, kv.Value))
-			} else if kv.SetFile {
-				if setFileOverridesBldr.Len() > 0 {
-					setFileOverridesBldr.WriteString(",")
+	// Expand the existing kvs values into expected format
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:422]: Is this trying to create one helm override file for all overrides in the CR?

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 14]
```diff
@@ -61,6 +64,9 @@ type HelmComponent struct {
 	// AppendOverridesFunc is an optional function get additional override values
 	AppendOverridesFunc appendOverridesSig
 
+	// GetHelmValueOverrides is an optional function that returns the helm override values
+	GetHelmValueOverrides getHelmValueOverridesSig
```
[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:68]: No. It's functionality is separate. It would be possible to combine the two if all of the helm components uptake this change.

**Review by fredtibbitts — COMMENTED** [review]
I don't pretend to understand all of the changes but they seem OK to me. Since the bulk of the changes are in the common helm component code, any reason not roll this feature out to all of the components? Makes sense to limit it in this PR but will we add it everywhere for 1.3?

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go (hunk 13 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:422]: Well, maybe I misunderstood your question. There will be one file per helm override source (configmap or secret)

File: platform-operator/controllers/verrazzano/component/certmanager/certmanager.go [hunk 15]
```diff
@@ -351,6 +354,23 @@ func isOCIDNS(vz *vzapi.Verrazzano) bool {
 
 // AppendOverrides Build the set of cert-manager overrides for the helm install
 func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ string, kvs []bom.KeyValue) ([]bom.KeyValue, error) {
+	// use image value for arg override
+	bomFile, err := bom.NewBom(config.GetDefaultBOMFilePath())
+	if err != nil {
+		return kvs, compContext.Log().ErrorNewErr("Failed to get the BOM file for the cert-manager image overrides: ", err)
+	}
+
+	images, err := bomFile.BuildImageOverrides("cert-manager")
```
[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:363]: This was done to piece together the image argument that had to change. Now that files are created, I had to change the image source for the component and update the value here.

[the remark this answers is not in the document]
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/certmanager/certmanager.go:363]: The rationale for this change is located here: https://github.com/verrazzano/verrazzano/pull/2967/files/24d017286a522a4bb15d063c2ffb024a8f92d1dc#r859763979

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go (hunk 12 above)
Comment (davisgomes) [reply platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:761]: Yea, eventually. This change was focused primarily on the Helm components
